adapter-base-quarkus/src/main/java/org/eclipse/hono/adapter/quarkus/AbstractProtocolAdapterApplication.java

```diff
@@ -472,4 +484,49 @@ private ClientConfigProperties commandResponseSenderConfig() {
         return props;
     }
 
+    /**
+     * Creates resource limit checks based on data retrieved from a Prometheus server
+     * via its HTTP API.
+     *
+     * @param config The configuration properties.
+     * @param tenantClient The client to use for retrieving tenant configuration data.
+     * @throws NullPointerException if any of the parameters are {@code null}-
+     * @return The checks.
+     */
+    protected ResourceLimitChecks prometheusResourceLimitChecks(
+            final PrometheusBasedResourceLimitChecksConfig config,
+            final TenantClient tenantClient) {
+
+        Objects.requireNonNull(config);
+        Objects.requireNonNull(tenantClient);
+
+        if (config.isHostConfigured()) {
+            final WebClientOptions webClientOptions = new WebClientOptions();
+            webClientOptions.setConnectTimeout(config.getConnectTimeout());
+            webClientOptions.setDefaultHost(config.getHost());
+            webClientOptions.setDefaultPort(config.getPort());
+            webClientOptions.setTrustOptions(config.getTrustOptions());
+            webClientOptions.setKeyCertOptions(config.getKeyCertOptions());
+            webClientOptions.setSsl(config.isTlsEnabled());
+
+            final var webClient = WebClient.create(vertx, webClientOptions);
+            final var cacheTimeout = Duration.ofSeconds(config.getCacheTimeout());
+            final Caffeine<Object, Object> builder = Caffeine.newBuilder()
+                    // make sure we run one Prometheus query at a time
+                    .executor(Executors.newSingleThreadExecutor())
+                    .initialCapacity(config.getCacheMinSize())
+                    .maximumSize(config.getCacheMaxSize())
+                    .expireAfterWrite(cacheTimeout)
+                    .refreshAfterWrite(cacheTimeout);
```

Comment:
This will result in expiration still taking precedence, because it indicates that the entry is invalid to return. On the other hand, a refresh is when the entry is stale but still acceptable to use. The two are meant to be used together so that active entries are reloaded and hide the latency, while inactive ones are allowed to fade away.

This data freshness is the same as a loaf of bread used for lunch:
1. just backed: enjoy!
2. stale: eat & order for delivery
3. moldy: throw away
  a) having a sandwich? go buy break and be hungry
  b) switching to a salad? no reason to keep or get more

You might therefore use `cacheTimeout / 2` or similar as a threshold. If a read request occurs within the refresh interval then it will be reloaded asynchronously. But if the entry has passed the expiration threshold, the cache won't serve that entry and penalize the caller as if it was absent.

You can use `Caffeine.ticker(t)` to control logical time, and use Guava's `FakeTicker` in your tests. That would allow you to write a test case demonstrating those different scenarios and observe Caffeine's behavior.

Reply:
Got it. Thanks for the hint :-) May I ask how you became aware of our using of Caffeine in Hono?

Reply:
I use GitHub’s global search to monitor recent activity. That way I can try to catch misunderstandings early before users get frustrated and also get feedback on where improvements are needed. I saw this PR and skimming it caught this setting.